### PR TITLE
add "admission" API group

### PR DESCRIPTION
**Pull request by whitlockjc**
This commit is an initial pass at providing an admission API group.
The API group is required by the webhook admission controller being
developed as part of https://github.com/kubernetes/community/pull/132
and could be used more as that proposal comes to fruition.

**Note:** This PR was created by following the [Adding an API Group](https://github.com/kubernetes/community/blob/master/contributors/devel/adding-an-APIGroup.md) documentation.

cc @smarterclayton

### CODE REVIEW

File: pkg/apis/admission/doc.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (sttts) [inline pkg/apis/admission/doc.go]: 2017

File: pkg/apis/admission/register.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// Resource takes an unqualified resource and returns a Group qualified GroupResource
+func Resource(resource string) schema.GroupResource {
+	return SchemeGroupVersion.WithResource(resource).GroupResource()
+}
+
+var (
+	// SchemeBuilder the schema builder
+	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
```
Comment (sttts) [inline pkg/apis/admission/register.go]: There was a recent change in all kube apigroups to use `LocalSchemeBuilder`, compare e.g. pkg/apis/rbac/v1alpha1/register.go. I guess we need that here as well? /cc @caesarxuchao

File: pkg/apis/admission/register.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	// AddToScheme handler to add items to the schema
+	AddToScheme = SchemeBuilder.AddToScheme
+)
+
+func addKnownTypes(scheme *runtime.Scheme) error {
+	scheme.AddKnownTypes(SchemeGroupVersion,
+		&AdmittanceReview{},
+	)
+	//	metav1.AddToGroupVersion(scheme, SchemeGroupVersion)
```
Comment (sttts) [inline pkg/apis/admission/register.go]: this line can go in the internal version.

File: pkg/apis/admission/v1alpha1/doc.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package,register
+// +k8s:conversion-gen=k8s.io/kubernetes/pkg/apis/admission
+// +k8s:openapi-gen=true
+// +k8s:defaulter-gen=TypeMeta
+
+// +groupName=imagepolicy.k8s.io
```
Comment (sttts) [inline pkg/apis/admission/v1alpha1/doc.go]: wrong apigroup

File: pkg/apis/admission/register.go (hunk 2 above)
Comment (sttts) [inline pkg/apis/admission/register.go]: Disregard my comment. It's not necessary in the internal version.

File: pkg/apis/admission/v1alpha1/types.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	// Allowed indicates whether or not the admission request was permitted.
+	Allowed bool `json:"allowed" protobuf:"varint,1,opt,name=allowed"`
+	// Reason should be empty unless Allowed is false in which case it may contain a short description of what is wrong.
+	// Kubernetes may truncate excessively long errors when displaying to the user.
+	// +optional
+	Reason string `json:"reason,omitempty" protobuf:"bytes,2,opt,name=reason"`
+}
+
+// NewAdmittanceReview returns an AdmittanceReview for the provided admission.Attributes
+func NewAdmittanceReview(attr admission.Attributes) AdmittanceReview {
```
Comment (sttts) [inline pkg/apis/admission/v1alpha1/types.go]: can we move this into a helper.go or something like that?

File: pkg/apis/admission/v1alpha1/types.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/pkg/apis/authentication/v1beta1"
```
Comment (sttts) [inline pkg/apis/admission/v1alpha1/types.go]: import as authenticationv1beta1

**Comment by whitlockjc** [comment]
Thanks @sttts.  I realized a couple of these when I ran `make test`, unfortunately it was after the PR was created.  I apologize.  The new commit should address these.

**Comment by whitlockjc** [comment]
Changes completed.  But...when running `make test`, I'm seeing this:

```
panic: Double registration of different types for admission.k8s.io/__internal, Kind=WatchEvent: old=k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1.InternalEvent, new=k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/met
a/v1.WatchEvent
```

I'm looking into it but if you have any ideas, please let me know.

**Comment by whitlockjc** [comment]
I've fixed a few more tests and now I'm left with two things:

1. How do I avoid generating the REST API and related stuff like OpenAPI/Swagger stuff?
2. I have one more test failure:

```
--- FAIL: TestCommonKindsRegistered (0.00s)
        serialization_test.go:174: no kind "ListOptions" is registered for version "admission/v1alpha1"
        serialization_test.go:179: expected admission/v1alpha1, Kind=ListOptions: admission/v1alpha1, Kind=ListOptions no kind "ListOptions" is registered for version "admission/v1alpha1"
        serialization_test.go:183: expected admission/v1alpha1, Kind=ListOptions: expected pointer, but got invalid kind
        serialization_test.go:174: no kind "DeleteOptions" is registered for version "admission/v1alpha1"
        serialization_test.go:179: expected admission/v1alpha1, Kind=DeleteOptions: admission/v1alpha1, Kind=DeleteOptions no kind "DeleteOptions" is registered for version "admission/v1alpha1"
        serialization_test.go:183: expected admission/v1alpha1, Kind=DeleteOptions: expected pointer, but got invalid kind
        serialization_test.go:174: no kind "ExportOptions" is registered for version "admission/v1alpha1"
        serialization_test.go:179: expected admission/v1alpha1, Kind=ExportOptions: admission/v1alpha1, Kind=ExportOptions no kind "ExportOptions" is registered for version "admission/v1alpha1"
        serialization_test.go:183: expected admission/v1alpha1, Kind=ExportOptions: expected pointer, but got invalid kind
FAIL
FAIL    k8s.io/kubernetes/pkg/api       4.475s
```

**Comment by sttts** [comment]
@whitlockjc where does the string `admission/v1alpha1` come from? Shouldn't it be `admission.k8s.io/v1alpha1`?

File: pkg/apis/admission/register.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/apis/admission/register.go]: Yeah, the internal version doesn't matter.

**Comment by whitlockjc** [comment]
Good catch, I found I used the wrong `<group>/<version>` in `hack/lib/init.sh`.  PR being updated shortly.

**Comment by whitlockjc** [comment]
No failing tests via `make test` related the changes in this PR.  I'm still not sure how to make sure that `types.go` is right so that it **DOES NOT** build a REST API.

**Comment by whitlockjc** [comment]
I think I need to remove the annotations from `types.go` that are related to generating a client.  Since I'm not 100% sure, and I've not found the information to help just yet, I'll leave it as-is.

**Comment by sttts** [comment]
If this is a webhook-only API, it should be enough to register the types in the/a scheme, without going via the registered/announce mechanism. In other words, we should get rid of the install.go.

For the same reason, the `metav1.ObjectMeta` in `AdmittanceReview` doesn't make sense as these are not API objects in the sense that they show up in a CRUD api endpoint on the apiserver.

**Comment by whitlockjc** [comment]
Thanks @sttts for explaining this stuff.

**Comment by whitlockjc** [comment]
Requested changes made.

**Comment by whitlockjc** [comment]
cc @lavalamp

File: pkg/client/listers/admission/internalversion/admittancereview.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	admission "k8s.io/kubernetes/pkg/apis/admission"
+)
+
+// AdmittanceReviewLister helps list AdmittanceReviews.
+type AdmittanceReviewLister interface {
```
Comment (sttts) [inline pkg/client/listers/admission/internalversion/admittancereview.go]: Are the listers a left-over from when we had ObjectMeta? If I read https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/lister-gen/generators/lister.go#L82 correctly, these shouldn't be generated.

**Comment by sttts** [comment]
@k8s-bot ok to test

File: pkg/client/listers/admission/internalversion/admittancereview.go (hunk 7 above)
Comment (whitlockjc) [inline pkg/client/listers/admission/internalversion/admittancereview.go]: It's likely left over.  I'll blow them away, rebuild and update the PR if they are no longer created.

**Comment by whitlockjc** [comment]
Just updated the commit to remove the lister stuff.  One thing I'm still not certain on is committing the Swagger stuff.  Since this is not a public REST API, there is no benefit to any Swagger artifacts being created/updated for `admission.k8s.io`.  Thoughts?

**Comment by whitlockjc** [comment]
I'm in the process of running `make test` and `make verify` to fix the remaining issues, even though the failing tests above don't seem related _(as of this moment, that may change)_.

**Comment by whitlockjc** [comment]
The last commit fixed `hack/verify-api-groups.sh` failing.  `hack/verify-pkg-names.sh` is failing because `hack/update-codecgen.sh` generates a package with an underscore to resolve a conflict where two separate packages end in `admission`.  I can work around this if need be but it's really a tooling bug if you ask me.  As for the failing openapi/swagger verifications, I'm looking into those next.

**Comment by whitlockjc** [comment]
Swagger verification also seem to be unrelated to my changes.  Running `hack/update-*.sh` _(as related to OpenAPI and Swagger)_ generates new files where some property names change case but that's it.

**Comment by whitlockjc** [comment]
The swagger changes being generated and failing the verification _might_ be related to the changes to `staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/group_version.go` which were required for other verification _(`json` tag required for external types)_.  I'm checking this now.

**Comment by whitlockjc** [comment]
Yup, that's it.  The reason the openapi/swagger verification fails is because of the changes to `staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/group_version.go`.  I remember making this change due to some test requiring it but that might had changed now that we're not exposing these types publicly.  I'll update the PR with my findings.

**Comment by whitlockjc** [comment]
As of the last update, all tests and verifications are working except for `hack/verify-pkg-names.sh` and that is because `pkg/apis/admission/v1alpha1/zz_generated.conversion.go` ends up with `pkg_admission` as a package alias due to both `k8s.io/apiserver/pkg/admission` and `k8s.io/kubernetes/pkg/apis/admission` being used.  It seems like if this is a bug in the `codecgen` stuff.

File: pkg/master/import_known_versions_test.go [hunk 8]
```diff
@@ -71,18 +71,20 @@ func TestTypeTags(t *testing.T) {
...
+	reflect.TypeOf(metav1.ListMeta{}):             true,
+	reflect.TypeOf(metav1.ObjectMeta{}):           true,
+	reflect.TypeOf(metav1.OwnerReference{}):       true,
+	reflect.TypeOf(metav1.LabelSelector{}):        true,
+	reflect.TypeOf(metav1.GetOptions{}):           true,
+	reflect.TypeOf(metav1.ExportOptions{}):        true,
+	reflect.TypeOf(metav1.ListOptions{}):          true,
+	reflect.TypeOf(metav1.DeleteOptions{}):        true,
+	reflect.TypeOf(metav1.GroupVersionKind{}):     true,
+	reflect.TypeOf(metav1.GroupVersionResource{}): true,
```
Comment (sttts) [inline pkg/master/import_known_versions_test.go]: why this change?

Comment (sttts) [inline pkg/master/import_known_versions_test.go]: I believe this makes sense. Disregard my question.

**Comment by sttts** [comment]
@whitlockjc can you add an exclude for `zz_generated.*.go` to hack/verify-pkg-names.sh. For bikeshedding about the `_` in imports, Daniel left this comment: `vendor/k8s.io/gengo/generator/import_tracker.go:42`. I think an exclude is fine.

File: pkg/master/import_known_versions_test.go (hunk 8 above)
Comment (whitlockjc) [inline pkg/master/import_known_versions_test.go]: Since `AdmittanceReviewSpec` has a struct field of `GroupVersionKind` and `GroupVersionResource`, it was required.  Now, I'm not 100% sure this is required anymore since the similar requirement for these types to have `json` tags is no longer required but at one point it was, likely when we were using `admission/install` and the API tests were firing.

**Comment by whitlockjc** [comment]
@sttts Exclude added to `hack/verify-pkg-names.sh` and rebased.  I'm going to run the tests and such again just in case but it should be good.  I'm also going to see if your [previous comment](https://github.com/kubernetes/kubernetes/pull/45564#discussion_r116418342) is valid now that we're not using `admission/install`.

**Comment by sttts** [comment]
@smarterclayton is this approved?

File: pkg/apis/admission/types.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/kubernetes/pkg/apis/authentication"
+)
+
+// AdmittanceReview describes an admission request.
+type AdmittanceReview struct {
```
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: I don't like `Admittance`.  `AdmissionReview` is more correct without bringing in an unusual word.

Comment (whitlockjc) [inline pkg/apis/admission/types.go]: I had that and lint warned me of package shadowing.  I can switch it back and just ignore the warning, I think it was only in the editor anyways.

File: pkg/apis/admission/types.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+// AdmittanceReviewSpec describes the admission.Attributes for the admission request.
+type AdmittanceReviewSpec struct {
+	// Kind is the type of object being manipulated.  For example: Pod
+	Kind metav1.GroupVersionKind
+	// Name is the name of the object as presented in the request.  On a CREATE operation, the client may omit name and
+	// rely on the server to generate the name.  If that is the case, this method will return the empty string.
+	Name string
+	// Namespace is the namespace associated with the request (if any).
+	Namespace string
+	// Object is the object from the incoming request prior to default values being applied
```
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: Note on this that the server is free to omit this value when returning a response to a caller.  That avoids a plugin having to serialize the object back, and the cost to decode from the response.

File: pkg/apis/admission/types.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+	Object runtime.Object
+	// OldObject is the existing object. Only populated for UPDATE requests.
```
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: Note on this that the server is free to omit this value when returning a response to a caller.  That avoids a plugin having to serialize the object back, and the cost to decode from the response.

File: pkg/apis/admission/types.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/kubernetes/pkg/apis/authentication"
+)
+
+// AdmittanceReview describes an admission request.
+type AdmittanceReview struct {
+	metav1.TypeMeta
+
+	// Spec describes the attributes for the admission request.
```
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: I would note that integrators implementing AdmissionReview are not required to populate `spec` in their response, or something to that effect.

File: pkg/apis/admission/types.go [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+	UserInfo authentication.UserInfo
+}
+
+// AdmittanceReviewStatus describes the status of the admission request.
+type AdmittanceReviewStatus struct {
+	// Allowed indicates whether or not the admission request was permitted.
+	Allowed bool
+	// Reason should be empty unless Allowed is false in which case it may contain a short description of what is wrong.
+	// Kubernetes may truncate excessively long errors when displaying to the user.
+	Reason string
```
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: This should be Message, not Reason (by description).  You want to have both `Message` and `Reason`, and use the same descriptions from `metav1.Status` (i.e. Reason is a constant, machine readable, Message is a description, human readable).  And preserve the part about truncation, that is important.

**Review by smarterclayton — CHANGES_REQUESTED** [review]
Don't like the name, some other comments I'd like added, otherwise is very close.

File: pkg/apis/admission/types.go (hunk 10 above)
Comment (whitlockjc) [inline pkg/apis/admission/types.go]: Will do.

File: pkg/apis/admission/types.go (hunk 12 above)
Comment (whitlockjc) [inline pkg/apis/admission/types.go]: Will do.

File: pkg/apis/admission/types.go (hunk 13 above)
Comment (whitlockjc) [inline pkg/apis/admission/types.go]: Will do.

**Comment by whitlockjc** [comment]
Feedback addressed.

**Comment by whitlockjc** [comment]
I'm currently moving the _note_ from the internal `types.go` to the external `types.go` with regard to `AdmissionReview.Spec` needing not be set in the webhook response.  Should have no real impact on anything else.

File: pkg/apis/admission/types.go [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/kubernetes/pkg/apis/authentication"
+)
+
+// AdmissionReview describes an admission request.
+type AdmissionReview struct {
+	metav1.TypeMeta
+
+	// Spec describes the attributes for the admission request.  (Note: Since this admission controller is non-mutating
+	// the webhook should avoid setting this in its response to avoid the cost of deserializing it.)
```
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: One slight tweak, instead of "should" use "may".  You can probably drop the parethensis and just do a second sentence.  

> Since admission controllers are primarily responsible for making decisions in the critical path of a request, implementors may omit the spec field from their response to minimize time spent serializing requests.  Callers must not rely on the spec field being preserved in a response.

File: pkg/apis/admission/types.go (hunk 11 above)
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: Not about mutating or not, just because of performance.  The mutating field would be returned in status.  That way a caller only has to check allowed, and optionally a response, if and only if it was changed.

File: pkg/apis/admission/types.go (hunk 14 above)
Comment (whitlockjc) [inline pkg/apis/admission/types.go]: Changes incoming.

**Comment by whitlockjc** [comment]
PR has been updated with changes reflecting the last round of feedback.  Thanks for your time and guidance @smarterclayton.

File: pkg/apis/admission/types.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+// AdmissionReviewStatus describes the status of the admission request.
+type AdmissionReviewStatus struct {
+	// Allowed indicates whether or not the admission request was permitted.
+	Allowed bool
+	// A human-readable description of the status of this operation.  Kubernetes may truncate excessively long errors when
+	// displaying to the user.
+	Message string
+	// A machine-readable description of why this operation is in the state it is. If this value is empty there is no
+	// information available.  A Reason clarifies an HTTP status code but does not override it.
+	Reason metav1.StatusReason
```
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: Oops, I meant status condition reason, not this reason.  See NodeConditions or PodConditions and the Reason and Message fields on them.

Comment (smarterclayton) [inline pkg/apis/admission/types.go]: Actually... hrm.  Design wise, we need to be able to effectively serialize an entire Status (that's what the structured errors are).

So this actually needs to return a Status object (nested) or an error.  The absence of a status could mean accepted, or we could explicitly require success.

Hrm.

Comment (smarterclayton) [inline pkg/apis/admission/types.go]: Admission is basically `func(Attributes) error`.  If `spec` is attributes, then this (status) needs to be a `metav1.Status`, probably modeled as:

```
type AdmissionReviewStatus  struct {
  Allowed bool
  // Reason is set if the webhook wishes to return additional information about the result of this operation. If Allowed is true, this field is informational only. If Allowed is false, this can alter the error an end user sees.
  Reason *metav1.Status  
}
```

I'm torn on `Reason` as the field name - I'm somewhat concerned with stuttering because Status already contains fields that match all possible names, i.e.:

* status.status.reason (redundant)
* status.reason.reason (redundant)
* status.details.details (redundant)
* status.error.details (we don't use "error" anywhere in the API today, and I'm hesitant to make the status field, which can represent success as well, also be negative)

Comment (whitlockjc) [inline pkg/apis/admission/types.go]: I agree.  I modeled it originally after the `ImagePolicy` webhook.  Any reason not to make `metav1.Status` inline?

Comment (whitlockjc) [inline pkg/apis/admission/types.go]: The more I think about it, I think the reasons you don't like `Reason *metav1.Status` is likely the reason [imagepolicy](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/imagepolicy/types.go#L59) did things the way it did.

Comment (smarterclayton) [inline pkg/apis/admission/types.go]: I would prefer not to inline it because we will add new fields to status for returned object and I'm not 100% sure we won't ever have a conflict.

Comment (smarterclayton) [inline pkg/apis/admission/types.go]: We can fairly easily rename fields, but inlining or non-inlining is really hard to undo.  Also, this will be alpha anyway so we have time to change.

Comment (smarterclayton) [inline pkg/apis/admission/types.go]: How about `Result`?  I'm using that in initializers, and it's pretty clear.

Comment (whitlockjc) [inline pkg/apis/admission/types.go]: But isn't the `Allowed` the _result_ and `Status` was merely to relay erroneous data back to the AC when there is an issue within the webhook.  Why not just make `AdmissionRequest.Status` a `metav1.Status` and figure out the best way to describe `yes` and `no` from that structure?

Comment (smarterclayton) [inline pkg/apis/admission/types.go]: We'll need to add new fields eventually into AdmissionRequest.Status that we wouldn't add to metav1.Status.

**Comment by whitlockjc** [comment]
I've updated the PR to use `AdmissionReviewStatus.Result` as a `metav1.Status` as requested.

File: pkg/apis/admission/types.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// AdmissionReviewStatus describes the status of the admission request.
+type AdmissionReviewStatus struct {
+	// Allowed indicates whether or not the admission request was permitted.
+	Allowed bool
+	// Result contains extra details into why an admission request was denied.
+	// This field IS NOT consulted in any way if "Allowed" is "true".
+	// +optional
+	Result metav1.Status
```
Comment (smarterclayton) [inline pkg/apis/admission/types.go]: Because this is optional it should be *metav1.Status internal and external.  Because we don't want callers to *have* to set it, and if it's not set we don't have to even deserialize it (this will be fairly perf critical code path as well)

Comment (whitlockjc) [inline pkg/apis/admission/types.go]: Ah right, that was an oversight.

**Comment by smarterclayton** [comment]
LGTM after that change and green

**Comment by whitlockjc** [comment]
@smarterclayton The PR has been updated as requested and the build appears green.  Let me know if there is more to do and thanks again for your help.

**Comment by whitlockjc** [comment]
@smarterclayton I'm not seeing how this commit could cause the failed _Jenkins verification_ job.  I'll keep digging.

**Comment by smarterclayton** [comment]
Regenerate protobuf

**Comment by whitlockjc** [comment]
I just tried and no changes are created.  I blew away all generated sources, updated codecgen, updated protobuf, updated bazel.

**Comment by whitlockjc** [comment]
Steps ran:

* `rm pkg/apis/admission/BUILD pkg/apis/admission/zz_generated.deepcopy.go pkg/apis/admission/v1alpha1/BUILD pkg/apis/admission/v1alpha1/zz_generated.* pkg/apis/admission/v1alpha1/generated.p*`
* `echo "package admission" > pkg/apis/admission/types.generated.go`
* `echo "package v1alpha1" > pkg/apis/admission/v1alpha1/types.generated.go`
* `./hack/update-codecgen.sh`
* `./hack/update-generated-protobuf.sh`
* `./hack/update-bazel.sh`

When I do this, I end up with a clean working directory.

**Comment by whitlockjc** [comment]
I had to rebase again.  PR updated with hopefully fixed protobufs.

**Comment by whitlockjc** [comment]
I'm not seeing where the failing _Jenkins verification_ is a result of this PR.  Any extra eyes would be appreciated.

**Comment by whitlockjc** [comment]
Found it, fix coming in.

**Comment by smarterclayton** [comment]
@k8s-bot verify test this

File: pkg/apis/admission/BUILD [hunk 17]
```diff
@@ -0,0 +1,46 @@
+package(default_visibility = ["//visibility:public"])
```
Comment (wojtek-t) [inline pkg/apis/admission/BUILD]: @whitlockjc For the future, please try to separate autogenerated changes into a separate commit - it's much easier to review then.

Comment (whitlockjc) [inline pkg/apis/admission/BUILD]: Are we sure this is advisable?  If you run `./hack/update-all.sh` as suggested in the [documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/adding-an-APIGroup.md), this is created.  Also, for new people it would be hard to distinguish between the different _types_ of autogenerated code, like bazel vs. codecgen vs. protobuf vs. openapi/swagger.  Following the documentation has all of these resources being created.

**Comment by smarterclayton** [comment]
You don't have to separate by type.  I often do "api" and then "generated:
api"

File: pkg/apis/admission/BUILD (hunk 17 above)
Comment (wojtek-t) [inline pkg/apis/admission/BUILD]: Yeah - as @smarterclayton wrote - you don't need to split it by type. It's fine to have all auto-generated things in a single commit.

**Comment by whitlockjc** [comment]
Okay, I got you.  So one commit with the API plumbing/types and one with the generated stuff.  Makes sense.  Is this documented somewhere, this desire to break this stuff up?  If not, I'd love to put this somewhere.
